src/pybind/mgr/cephadm/module.py

```diff
@@ -1859,6 +1859,8 @@ def matches_network(host):
                 daemon_id=daemon_id,
             )
             daemons.append(sd)
+            if daemon_type in ['grafana', 'iscsi', 'prometheus', 'alertmanager', 'nfs']:
+                self._get_cephadm_service(daemon_type).daemon_check_post([sd])
```

Comment:
instead of having this in `_apply_service`, what do you think of moving this into _create_daemon ?

Reply:
I think the reason I chose to put it in _apply_service over _create_daemon originally was because the redeploy  command also goes through _create_daemon and I didn't want to alter the url in that case. If you think changing the url is also okay in that case then I could move it there.

Reply:
seems like we should update the dashboard urls during redeploy?

NFS pool/namespace could change during redeploy--
https://github.com/ceph/ceph/blob/master/src/pybind/mgr/cephadm/services/nfs.py#L74-L95

Similar for iscsi too --
https://github.com/ceph/ceph/blob/master/src/pybind/mgr/cephadm/services/iscsi.py#L64-L90

Reply:
Okay, in that case I will move it to _create_daemon where it should update during redeploy as well as when the placement changes

Reply:
Completely against about this change. ( overwrite when redeploy).
If we are redeploying the daemon in the same host...  (it does not matter what daemon) probably this is because we want to upgrade the container image or restart the daemon. 
Overwriting config values with default values is something that probably is not going to make very happy to the user that has change the settings.

I think that it is ok to use default or calculated values when you are deploying for first time a container in one host... but if it has been deployed .. there is a true possibility that the user had changed settings to customize the daemon behaviour. Overwriting config settings is always dangerous.... so my suggestion is not to do this kind of things. (in any daemon)

Reply:
> Overwriting config settings is always dangerous.... so my suggestion is not to do this kind of things. (in any daemon)

Hence my suggestion to **never** overwrite any dashboard settings, but instead **only** provide default values for the dashboard.

Under the assumption that we go withthe solution to only overwrite the dashboard settings **sometimes** (like when deploying services, failovers etc), we **have** provide a way for users to "fix" the dashboard config, thus :+1: for also doing this on redeploy.

Reply:
Just an update on this. I've been trying to do the check post call within _create_daemon but have run into a problem where the mon commands it sends in the process (it sends a "dashboard get-grafana-api-url" and "dashboard set-grafana-apit url <new url>" both within the _check_and_set_dashboard function of services/cephadmservice.py) never returns. Furthermore, if I interrupt  the redeploy to regain control and then try to manually get or set the grafana url afterwards those commands also don't return. I haven't been able to figure out why this is happening. For reference, I only intended to add the two highlighted lines here to the _create_daemon function. I'm still looking into it but I haven't had much luck.
![image](https://user-images.githubusercontent.com/47704447/85807258-55d34600-b71f-11ea-8ef3-c32412959cef.png)

Reply:
also, this seems to only be an issue during redeploy. _create_daemon runs fine with the daemon_check_post call during initial deployment and when changing placement.

Reply:
I don't think a direct call within _create_daemon is possible due to this https://docs.ceph.com/docs/master/dev/cephadm/#note-regarding-network-calls-from-cli-handlers
so I've tried this new method of setting a sort of flag within _create_daemon so the post actions can be handled after within the service loop to hopefully get the same outcome as doing the post actions in _create_daemon without the mon commands hanging